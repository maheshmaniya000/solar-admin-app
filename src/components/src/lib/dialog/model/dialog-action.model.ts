import { Observable } from 'rxjs';

export interface DialogAction<T = any> {
  dismissing: boolean;
  textTranslationKey: string;
  payload: T;
  disabled?: boolean;
  dataTest: string;
}

// https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-1.html#mapped-types-on-tuples-and-arrays
type TypesToDialogActions<Types extends any[]> = { [K in keyof Types]: DialogAction<Types[K]> };

export type DialogActions<PayloadTypes extends any[] = any[]> =
  | TypesToDialogActions<PayloadTypes>
  | Observable<TypesToDialogActions<PayloadTypes>>;
