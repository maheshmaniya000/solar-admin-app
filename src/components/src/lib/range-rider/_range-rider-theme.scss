@use '~@angular/material' as mat;

@mixin theme($theme-or-color-config) {
  $color-config: mat.get-color-config($theme-or-color-config);
  $primary: map-get($color-config, primary);
  $background: map-get($color-config, background);

  sg-range-rider {
    background-color: mat.get-color-from-palette($background, status-bar);

    .rider {
      border-color: mat.get-color-from-palette($primary);
    }
  }
}
