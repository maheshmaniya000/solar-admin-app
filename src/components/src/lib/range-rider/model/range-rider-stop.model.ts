export interface RangeRiderStop {
  value: number;
  color: string;
}
