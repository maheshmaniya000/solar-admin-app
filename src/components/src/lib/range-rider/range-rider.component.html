<div class="rider" [style.left.%]="riderLeftOffset"></div>
<div
  class="active-interval"
  [style.width.%]="activeIntervalConfig?.width"
  [style.left.%]="activeIntervalConfig?.leftOffset"
  [style.background]="activeIntervalConfig?.background"
></div>
