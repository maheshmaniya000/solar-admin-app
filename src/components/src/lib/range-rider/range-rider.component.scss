@use "sass:math";
@use 'variables';

$rider-width: 7px;
$rider-height: 5px;

:host {
  position: relative;
  display: inline-block;
  width: 100%;
  height: 6px;
}

.rider {
  position: absolute;
  bottom: 6px;
  width: 0;
  height: 0;

  border-style: solid;
  border-width: $rider-height math.div($rider-width, 2) 0 math.div($rider-width, 2);
  border-right-color: transparent;
  border-bottom: none;
  border-left-color: transparent;
  margin-left: -$rider-width;
  transition: variables.$sg-slider-transition;
}

.active-interval {
  position: absolute;
  height: 6px;
  transition: variables.$sg-slider-transition;
}
