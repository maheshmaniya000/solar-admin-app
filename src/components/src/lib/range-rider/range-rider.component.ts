import {
  ChangeDetectionStrategy,
  Component,
  Input,
  OnChanges
} from '@angular/core';
import { clamp, first, initial, inRange, isNil, last, sortBy, tail, zip } from 'lodash-es';

import { Range } from '../models/range.model';
import { calculatePercentageWithinRange } from '../utils';
import { RangeRiderStop } from './model/range-rider-stop.model';

@Component({
  selector: 'sg-range-rider',
  templateUrl: './range-rider.component.html',
  styleUrls: ['./range-rider.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class RangeRiderComponent implements OnChanges {
  @Input() stops: RangeRiderStop[];
  @Input() value: number;

  activeIntervalConfig: {
    width: number;
    leftOffset: number;
    background: string;
  };
  riderLeftOffset: number;

  private static calculateIntervalPercentageRange(
    interval: Range<RangeRiderStop>,
    totalRange: Range
  ): Range {
    return {
      min: calculatePercentageWithinRange(
        interval.min.value,
        totalRange.min,
        totalRange.max
      ),
      max: calculatePercentageWithinRange(
        interval.max.value,
        totalRange.min,
        totalRange.max
      )
    };
  }

  private static buildIntervalBackgroundLinearGradient(
    interval: Range<RangeRiderStop>
  ): string {
    return `linear-gradient(to right, ${interval.min.color}, 50%, ${interval.max.color})`;
  }

  ngOnChanges(): void {
    this.validateStops();
    const intervals = this.getIntervals();
    const totalRange = Range.from(
      first(intervals).min.value,
      last(intervals).max.value
    );
    const activeInterval = this.getActiveInterval(intervals, totalRange);
    const activeIntervalPercentageRange = RangeRiderComponent.calculateIntervalPercentageRange(
      activeInterval,
      totalRange
    );
    this.activeIntervalConfig = {
      width: Range.size(activeIntervalPercentageRange),
      leftOffset: activeIntervalPercentageRange.min,
      background: RangeRiderComponent.buildIntervalBackgroundLinearGradient(
        activeInterval
      )
    };
    this.riderLeftOffset = this.calculateRiderLeftOffset(totalRange);
  }

  private validateStops(): void {
    if (isNil(this.stops) || this.stops.length < 2) {
      throw new Error(
        `Minimum number of stops is 2, got ${this.stops?.length ?? undefined}`
      );
    }

    this.stops.forEach(stop => {
      const errorMessage = ` property of range rider stop object ${JSON.stringify(
        stop
      )} is undefined`;

      if (isNil(stop.value)) {
        throw new Error(`Value ${errorMessage}`);
      }

      if (isNil(stop.color)) {
        throw new Error(`Color ${errorMessage}`);
      }
    });
  }

  private getIntervals(): Range<RangeRiderStop>[] {
    const sortedStops = sortBy(this.stops, ({ value }) => value);
    return zip(initial(sortedStops), tail(sortedStops))
      .map(([from, to]) => Range.from(from, to))
      .filter(({ min, max }) => min.value !== max.value);
  }

  private getActiveInterval(
    intervals: Range<RangeRiderStop>[],
    range: Range
  ): Range<RangeRiderStop> {
    const clampedValue = clamp(this.value ?? 0, range.min, range.max);
    return intervals.find(
      ({ min, max }) =>
        inRange(clampedValue, min.value, max.value) ||
        clampedValue === max.value
    );
  }

  private calculateRiderLeftOffset(range: Range): number {
    return isNil(this.value)
      ? 0
      : calculatePercentageWithinRange(this.value, range.min, range.max);
  }
}
