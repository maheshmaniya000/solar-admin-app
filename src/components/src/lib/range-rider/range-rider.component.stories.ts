import { moduleMetadata, Story } from '@storybook/angular';

import { SgDefaultsModule } from '../defaults/defaults.module';
import { RangeRiderMocks } from './mock/range-rider.mocks';
import { RangeRiderModule } from './range-rider.module';

export default {
  title: 'Solargis/Components/Range Rider',
  decorators: [
    moduleMetadata({
      imports: [RangeRiderModule, SgDefaultsModule]
    })
  ]
};

export const predefinedStopsArrays: Story = props => ({
  props,
  template:
    '<sg-range-rider [stops]="stops" [value]="value"></sg-range-rider><br><br>Array of stops: <pre>{{stops | json}}</pre>'
});

predefinedStopsArrays.argTypes = {
  value: { name: 'Value', control: { type: 'number' }, defaultValue: 5 },
  stops: {
    name: 'Stops',
    control: {
      type: 'select',
      options: {
        simple: RangeRiderMocks.simpleMockStops,
        'with interval of one color':
          RangeRiderMocks.mockStopsWithIntervalOfOneColor
      }
    },
    defaultValue: RangeRiderMocks.simpleMockStops
  }
};

export const configurableStopsArray: Story = props => ({
  props,
  template: '<sg-range-rider [stops]="stops" [value]="value"></sg-range-rider>'
});

configurableStopsArray.argTypes = {
  stops: {
    name: 'Stops',
    control: { type: 'object' },
    defaultValue: RangeRiderMocks.mockStopsWithIntervalOfOneColor
  },
  value: {
    name: 'Value',
    control: { type: 'number' },
    defaultValue: 18
  }
};
