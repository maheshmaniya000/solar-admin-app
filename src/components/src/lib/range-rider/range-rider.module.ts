import { NgModule } from '@angular/core';

import { RangeRiderComponent } from './range-rider.component';

@NgModule({
  declarations: [RangeRiderComponent],
  exports: [RangeRiderComponent]
})
export class RangeRiderModule {}
