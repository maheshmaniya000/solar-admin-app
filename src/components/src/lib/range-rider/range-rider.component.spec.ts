import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';

import { RangeRiderMocks } from './mock/range-rider.mocks';
import { RangeRiderStop } from './model/range-rider-stop.model';
import { RangeRiderComponent } from './range-rider.component';

describe('RangeRiderComponent', () => {
  describe('unit', () => {
    let component: RangeRiderComponent;
    beforeEach(() => {
      component = new RangeRiderComponent();
    });

    describe('ngOnChanges', () => {
      ([
        [],
        undefined,
        null,
        [{ value: 5, color: '#FFF' }],
        [
          { value: undefined, color: '#FFF' },
          { value: 5, color: '#FFF' }
        ],
        [
          { value: null, color: '#FFF' },
          { value: 5, color: '#FFF' }
        ],
        [
          { value: 5, color: undefined },
          { value: 5, color: '#FFF' }
        ],
        [
          { value: 5, color: null },
          { value: 5, color: '#FFF' }
        ],
        [
          { value: 5, color: '#FFF' },
          { value: 5, color: '#FFF' }
        ]
      ] as RangeRiderStop[][]).forEach(stops => {
        it(`should throw when input '${JSON.stringify(
          stops
        )}' is not valid array of range rider stops`, () => {
          component.stops = stops;
          expect(() => component.ngOnChanges()).toThrow();
        });
      });

      describe('updated properties', () => {
        beforeEach(() => {
          component.stops = RangeRiderMocks.mockStopsWithIntervalOfOneColor;
        });

        const setupTest = (value: number): void => {
          component.value = value;
          component.ngOnChanges();
        };

        const buildLinearGradientExpression = (
          startColor: string,
          endColor: string
        ): string =>
          `linear-gradient(to right, ${startColor}, 50%, ${endColor})`;

        ([
          {
            testValue: undefined,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 0,
              background: buildLinearGradientExpression('#E30000', '#F4B300')
            },
            expectedRiderLeftOffset: 0
          },
          {
            testValue: null,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 0,
              background: buildLinearGradientExpression('#E30000', '#F4B300')
            },
            expectedRiderLeftOffset: 0
          },
          {
            testValue: -5,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 0,
              background: buildLinearGradientExpression('#E30000', '#F4B300')
            },
            expectedRiderLeftOffset: 0
          },
          {
            testValue: 0,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 0,
              background: buildLinearGradientExpression('#E30000', '#F4B300')
            },
            expectedRiderLeftOffset: 0
          },
          {
            testValue: 5,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 0,
              background: buildLinearGradientExpression('#E30000', '#F4B300')
            },
            expectedRiderLeftOffset: 0
          },
          {
            testValue: 4.99,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 0,
              background: buildLinearGradientExpression('#E30000', '#F4B300')
            },
            expectedRiderLeftOffset: 0
          },
          {
            testValue: 6,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 0,
              background: buildLinearGradientExpression('#E30000', '#F4B300')
            },
            expectedRiderLeftOffset: 4
          },
          {
            testValue: 15,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 0,
              background: buildLinearGradientExpression('#E30000', '#F4B300')
            },
            expectedRiderLeftOffset: 40
          },
          {
            testValue: 15.01,
            expectedActiveIntervalConfig: {
              width: 20,
              leftOffset: 40,
              background: buildLinearGradientExpression('#94C34A', '#94C34A')
            },
            expectedRiderLeftOffset: 40.04
          },
          {
            testValue: 20,
            expectedActiveIntervalConfig: {
              width: 20,
              leftOffset: 40,
              background: buildLinearGradientExpression('#94C34A', '#94C34A')
            },
            expectedRiderLeftOffset: 60
          },
          {
            testValue: 25.01,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 60,
              background: buildLinearGradientExpression('#F4B300', '#E30000')
            },
            expectedRiderLeftOffset: 80.04
          },
          {
            testValue: 30,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 60,
              background: buildLinearGradientExpression('#F4B300', '#E30000')
            },
            expectedRiderLeftOffset: 100
          },
          {
            testValue: 30.01,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 60,
              background: buildLinearGradientExpression('#F4B300', '#E30000')
            },
            expectedRiderLeftOffset: 100
          },
          {
            testValue: 35,
            expectedActiveIntervalConfig: {
              width: 40,
              leftOffset: 60,
              background: buildLinearGradientExpression('#F4B300', '#E30000')
            },
            expectedRiderLeftOffset: 100
          }
        ] as {
          testValue: number | undefined | null;
          expectedActiveIntervalConfig: {
            width: number;
            leftOffset: number;
            background: string;
          };
          expectedRiderLeftOffset: number;
        }[]).forEach(
          ({
            testValue,
            expectedActiveIntervalConfig,
            expectedRiderLeftOffset
          }) => {
            const testValueMessage = `; (test value: ${testValue})`;

            it(`should contain correct active interval config${testValueMessage}`, () => {
              setupTest(testValue);
              expect(component.activeIntervalConfig).toEqual(
                expectedActiveIntervalConfig
              );
            });

            it(`should contain correct left offset of rider div element in percentage${testValueMessage}`, () => {
              setupTest(testValue);
              expect(component.riderLeftOffset).toBe(expectedRiderLeftOffset);
            });
          }
        );
      });
    });
  });

  describe('component', () => {
    const createComponent = createComponentFactory({
      component: RangeRiderComponent,
      shallow: true
    });
    let spectator: Spectator<RangeRiderComponent>;

    beforeEach(() => {
      spectator = createComponent({
        props: {
          stops: RangeRiderMocks.mockStopsWithIntervalOfOneColor,
          value: 10
        }
      });
      spectator.component.ngOnChanges();
    });

    it('should render with correctly bound attributes', () => {
      expect(spectator.fixture).toMatchSnapshot();
    });
  });
});
