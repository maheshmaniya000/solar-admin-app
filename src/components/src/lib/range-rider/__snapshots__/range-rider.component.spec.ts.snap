// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeRiderComponent component should render with correctly bound attributes 1`] = `
<sg-range-rider
  __ngContext__={[Function LRootView]}
  activeIntervalConfig={[Function Object]}
  riderLeftOffset={[Function Number]}
  stops={[Function Array]}
  value={[Function Number]}
>
  <div
    class="rider"
    style="left: 20%;"
  /><div
    class="active-interval"
    style="width: 40%; left: 0%;"
  />
</sg-range-rider>
`;
